Automated Development Environment Setup Rake Tasks

charliescheer opened the pull request:
This PR adds new tasks to the WordPress iOS rakefile to make it easy for new internal or external developers to get setup with WPiOS quckly.

The new tasks, rake init:oss and rake init:developer, automate much of the setup process that is in the current read me file into a single command that can be run on a new download of WordPress to do the following:
Check for xcode is installed with the correct version
Install all dependencies
install additional development tools
install lint hooks
setup API credentials and generate a filled out wpcom_app_credentials file
And help internal developers create a gpg key for access to production credentials

I have also created a new version of the read me that reflects how to use these new scripts when getting started with WPiOS for the first time.  

To test:
1. Clone a copy of WordPress-ios into a new folder.
2. in terminal go to that directory and run `rake init:oss` or `rake init:developer`  
3. The scripts will run mostly without input 
4. Make sure to have a valid API client App ID and Client secret, you will be prompted to add this info
5.  A new gpg key will also be created in this process (there is an option to skip that)

Once complete make sure that:
The dependencies have been installed and are up to date
the swift link hooks are installed and setup
WordPress ios can compile and log in

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+  credentials:setup
+]
+
+desc "Install and configure WordPress iOS and it's dependencies - a8c Developers"
```
```suggestion
desc "Install and configure WordPress iOS and it's dependencies - Automattic Developers"
```

Before joining, I wasn't familiar with what A8C meant.

charliescheer replied:
Fixed

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+    namespace :xcode_app do
+      #check the existance of xcode, and compare version to CI specs
+      task :check do
+        puts "Checking for system for XCode"
```
**Nitpick**, it's called [Xcode](https://developer.apple.com/xcode/) 😆
```suggestion
        puts "Checking for system for Xcode"
```

charliescheer replied:
Eek found several of those... fixed

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+
+        puts "Checking CI recommendded installed XCode version"
+
+        if !xcode_version_is_correct?
```
**Ruby nitpick**, I find `!value` as a conditional tricky to read, it's easy to miss the `!` while scanning the code and getting the wrong idea of what the code does.

Possible alternatives are

```ruby
if xcode_version_is_correct == false
  ... 
end
```

or 

```ruby
unless xcode_version_is_correct
  ... 
end
```

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+          #if xcode is the wrong version, prompt user to install the correct version and terminate rake
+          puts "Not recommended version of XCode installed"
+          puts "It is recommended to use Xcode version #{get_ci_xcode_version}"
+          STDOUT.puts "Please press enter to continue"
```
Is this `STDOUT` required? How does this differ from the other `puts`?

charliescheer replied:
Ah that isn't required.

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+          puts "Not recommended version of XCode installed"
+          puts "It is recommended to use Xcode version #{get_ci_xcode_version}"
+          STDOUT.puts "Please press enter to continue"
+          complete = STDIN.gets.strip
```
**Nitpick**, it doesn't look like we need to read the value of `complete`, i.e., the key the user pressed. If so, we could just ignore it.

```suggestion
          STDIN.gets.strip
```

charliescheer replied:
Thanks for pointing that out! originally I was using this return and didn't remove it later.

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+      end
+
+      #compare xcode version to expected CI spec version
+      def xcode_version_is_correct?
+        if get_xcode_version == get_ci_xcode_version
+          puts "Correct version of XCode installed"
+          return true
+        end
+      end
```
Is it ok for this to return `nil` if the two versions don't match? Or would it be clearer if it returned false?

<img width="537" alt="Screen Shot 2020-05-27 at 3 22 40 pm" src="https://user-images.githubusercontent.com/1218433/82980922-07525f00-a02e-11ea-8719-f7397b475c7a.png">

charliescheer replied:
That's a good question.  I had not added any return to false because what it triggers happen if what comes back is not true.  So my thought on this was to only create the circumstance t return true, and left out other cases.  
But since it is a boolean check it would be clearer to return false if  the check fails.

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+        ci_config = File.read(".circleci/config.yml")
+        specs = YAML.load(ci_config)
+
+        ci_version = specs["jobs"]["Build Tests"]["executor"]["xcode-version"]
```
Nice idea going to check the Circle CI config for this value 👍 

I wonder if there's something that could be done to makes sure this code doesn't get out of date, say for example if the format in which Circle CI wants its YAML configs changes 🤔 

Just a question though, not something that should be addressed in this PR.

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+      task :check do
+        puts "Checking system for Homebrew"
+        unless command?("brew")
+          Rake::Task["tools:homebrew:install"].invoke
```
How would you feel about asking the user permission to run this?

charliescheer replied:
Great idea!  Added

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+        tool_check(developer_tools)
+      end
+
+      #check machine for if the developer tool is present, if not install
```
```suggestion
      #check if the developer tool is present in the machine, if not install
```

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+
+  namespace :lint do
+    task :check do
+      if !git_initialized?
+        puts "Initializing git repository"
+        sh "git init", verbose: false
+      end
```
Is this check necessary? Is there a scenario in which a user might run this code on the WordPress repo _without_ the `.git` folder set up? 🤔

charliescheer replied:
I hadn't put this in originally, but while testing, I setup a completely fresh install of WPiOS  (new OSX, new Xcode, new clone of the WPiOS repo) and the setup failed because I had not yet initialized the git repo on the folder.  So I thought I would build the check in just incase.

I agree that it is really unlikely that a user would run this setup without the .git folder having been created, but it the folder isn't there then the rake setup aborts.  

I do think that the check should be there, because if isn't and installing the swift lint hooks fails, there isn't a lot of indication as to why. But I could add some notes to clarify why the check is there.  What do you think?  Or is there a reason that should be left out that I missed?

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+  end
+
+  def prompt_for_continue
+    STDOUT.puts "Please press enter to continue"
```
**Nitpick** This text will be read by users that are expected to do something _and only then_ press enter to continue.

How can we make this clearer?

charliescheer replied:
Added some more details to make what the script is waiting for to the developer running it

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+  end
+
+  def get_client_id
+    STDOUT.puts "Please enter your Cliet id"
```
```suggestion
    STDOUT.puts "Please enter your Client ID"
```

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+    new_file = File.new(".configure-files/wpcom_app_credentials", "w")
+    File.open(".configure-files/temp_wpcom_app_credentials") do |file|
+      file.each_line do |line|
+        string = line.to_s()
```
Is this conversion to String necessary?

charliescheer replied:
huh.... could have sworn that I was getting errors before when the to_s wasn't there... but removed it and it is working now...  Simplified that section.  Thanks!!

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+  #prompt for a Y or N response, continue asking if other character
+  #return true for Y and false for N
+  def display_prompt_response
+    response = STDIN.gets.strip.upcase
```
Nice touch using the `.upcase` on the input 👍

charliescheer replied:
Thanks!

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+        response = STDIN.gets.strip.upcase
+    end
+
+    if response == "Y"
+      return true
+    elsif response == "N"
+      return false
+    end
```
**Ruby nitpick**, given we already have the logic above to ensure the only valid values are `Y` or `N`, this could be simplified into a check for `Y`

```suggestion
    return response == "Y"
```

charliescheer replied:
That is much simpler, thanks for the advice!  Fixed

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,399 @@ task :xcode => [:dependencies] do
+    puts "====================================================================================="
+  end
+
+#end namespace GPG_key_setup
+end
```
While I appreciate the information value these "end namespace" comments give when scanning the code, I do wonder whether the cost of maintaining them or them getting out of date (e.g., the `namespace` here is called `:gpg_key` not `:GPG_key_setup`) is too high.

charliescheer replied:
Great point, I hadn't considered that at the time.

mokagio commented on README.md:
```diff
@@ -7,28 +7,35 @@
+1. Now, run `cd WordPress-iOS` to enter the working directory.
 
-In order to login to WordPress.com using the app:
+### Create WordPress.com API Credentials
```
Using the same heading level here as above, `###`, made me think this was a different section, while it's a subsection of the setup instructions
```suggestion
#### Create WordPress.com API Credentials
```

mokagio commented on README.md:
```diff
@@ -7,28 +7,35 @@
-You can only log in with the WordPress.com account that you used to create the WordPress application.
+## Configuration Details
+
+The steps above will help you configure the WordPress app to run and compile.  But you may sometimes need to update or re run specific parts of the initial setup (like updating the dependencies)  To see how to do that, please check out the steps below
```
```suggestion
The steps above will help you configure the WordPress app to run and compile. But you may sometimes need to update or re-run specific parts of the initial setup (like updating the dependencies.) To see how to do that, please check out the steps below.
```

charliescheer commented:
Thanks for the great feedback @mokagio I have fixed the things that you suggested and added a check with the developer before installing homebrew.

Please let me know if you see anything else that needs to be looked at/fixed.

mokagio commented on Rakefile:
```diff
@@ -262,6 +263,397 @@ task :xcode => [:dependencies] do
+  end
+
+  def get_client_id
+    STDOUT.puts "Please enter your Client id"
```
Just a nitpick. You don't have to address this here, it an easily be a followup PR because it doesn't have to do with the actual workings of the code.

There's a bit of inconsistencies in how the "Client ID" and "Client Secret" are typed. Sometimes they are "Client ID", other times "client ID", or "Client id", or "client id". It would be great to have it consistent. To be honest, though, [our docs](https://developer.wordpress.com/docs/oauth2/) are not consistent with it either... 😳 

```suggestion
    STDOUT.puts "Please enter your Client ID"
```

charliescheer commented on Rakefile:
```diff
@@ -636,6 +624,18 @@ namespace :gpg_key do
   end
 end
 
+#prompt for a Y or N response, continue asking if other character
+#return true for Y and false for N
+def display_prompt_response
+  response = STDIN.gets.strip.upcase
+  until response == "Y" || response == "N"
+      puts "Invalid entry, please enter Y or N"
+      response = STDIN.gets.strip.upcase
+  end
+
+  return response == "Y"
+end
```
Thanks!

loremattei commented:
Thanks for this @charliescheer and @mokagio!
I've started the CI tests on this PR to make sure everything is fine as we use `rake` on CI.
